packages/katello/katello/hostname-change.rb

```diff
@@ -239,6 +281,116 @@ def setup_opt_parser
       @opt_parser.parse!
     end
 
+    def dns_managed?
+      @scenario_answers['foreman_proxy']['dns'] &&
+        @scenario_answers['foreman_proxy']['dns_managed']
+    end
+
+    def should_update_dns?
+      dns_managed? && !@options[:skip_dns]
+    end
+
+    def query_dns_server(fqdn, zone, local_ip)
+      query_dns = Resolv::DNS.new(nameserver: [local_ip], search: [zone], ndots: 1)
+      a_record = query_dns.getresource(fqdn, Resolv::DNS::Resource::IN::A)
+      ip = a_record.address.to_s if a_record
+      soa_record = query_dns.getresource(zone, Resolv::DNS::Resource::IN::SOA)
+      serial = soa_record.serial if soa_record
+      {
+        ip: ip,
+        serial: serial,
+      }
+    end
+
+    def new_dns_values
+      STDOUT.puts 'gathering DNS data...'
+      new_vals = OpenStruct.new
+      new_vals.local_ip = '127.0.0.1'
+      new_vals.zone = @scenario_answers['foreman_proxy']['dns_zone']
+      new_vals.reverse_zone = @scenario_answers['foreman_proxy']['dns_reverse']
+      if new_vals.reverse_zone.is_a?(Array)
+        new_vals.reverse_zone = new_vals.reverse_zone.first
+      end
+      new_vals.soa_admin_domain = "root.#{new_vals.zone}"
+      new_vals.key_file = @scenario_answers['foreman_proxy']['keyfile']
+      new_vals.old_fqdn = @old_hostname
+      new_vals.new_fqdn = @new_hostname
+
+      STDOUT.puts 'querying local DNS server...'
+      ip_serial_data = query_dns_server(@old_hostname, new_vals.zone, new_vals.local_ip)
+      new_vals.ip = ip_serial_data[:ip]
+      serial = ip_serial_data[:serial]
+      new_vals.new_serial = serial + 1 if serial
+      reverse_serial = query_dns_server(@old_hostname, new_vals.reverse_zone, new_vals.local_ip)[:serial]
+      new_vals.new_reverse_serial = reverse_serial + 1 if reverse_serial
+      run_cmd('rndc thaw')
+
+      new_vals
+    end
+
+    def nsupdate_command(dns_entries, key_file)
+      "echo -e \"#{dns_entries}\" | nsupdate -l -k #{key_file}"
+    end
+
+    def update_dns_records(d)
+      %w[local_ip zone reverse_zone soa_admin_domain key_file old_fqdn new_fqdn ip new_serial new_reverse_serial].each do |value|
+        unless d[value]
+          fail_with_message """
+  Error gathering DNS data: couldn't find value for '#{value}'.
+  Make sure /etc/foreman-installer/scenarios.d/#{@options[:scenario]}-answers.yaml
+  'foreman-proxy' section has values for dns_zone, dns_reverse, and keyfile.
+  Make sure A and SOA records are present for #{@old_hostname}.
+"""
+        end
+      end
+
+      STDOUT.puts 'updating DNS records:'
+      # Use nsupdate to update DNS records
+      # Update SOA record to new hostname; add new A and NS records; delete old A and NS records.
+      # Multi-line strings are not indented because Ruby 2.0 doesn't support <<~ heredocs
+
+      forward_dns_command = <<-HEREDOC
+local #{d.local_ip}
+zone #{d.zone}
+update add #{d.zone} 10800 SOA #{d.new_fqdn} #{d.soa_admin_domain}. #{d.new_serial} 86400 3600 604800 3600
```

Comment:
What if the admin changed values on the SOA?

Reply:
@ekohl The custom values would be overwritten.  I could collect and pass through the five additional values, but I feel like this specific case wouldn't happen often.  What about adding a blurb to the confirmation?
```
DNS records will be updated using nsupdate.  If you have any custom values in the SOA
 record, they will be overwritten.  If you'd rather update DNS records manually, re-run with the 
--skip-dns option.  Continue with changing hostname? [y/n]
```
or something like that